fhir/src/main/java/gov/cms/ab2d/fhir/IdentifierUtils.java

```diff
@@ -100,32 +161,113 @@ private static String getValue(ICompositeType identifier) {
         return (String) Versions.invokeGetMethod(identifier, "getValue");
     }
 
-    private static boolean isBeneficiaryId(ICompositeType identifier) {
-        String system = getSystem(identifier);
-        String value = getValue(identifier);
-        if (StringUtils.isAnyBlank(system, value)) {
-            return false;
+    /**
+     * From a FHIR identifier include MBI information, get the currency information
+     *
+     * @param identifier - the identifier
+     * @return - the currency information
+     */
+    private static PatientIdentifier.Currency getCurrencyMbi(ICompositeType identifier) {
+        Optional<IBaseExtension> currencyExtension = getCurrencyExtension(identifier);
+        if (currencyExtension.isEmpty()) {
+            return PatientIdentifier.Currency.UNKNOWN;
+        }
+        Object code = Versions.invokeGetMethod(currencyExtension.get(), "getValue");
+        String codeValue = (String) Versions.invokeGetMethod(code, "getCode");
+        switch (codeValue) {
+            case HISTORIC_MBI:
+                return PatientIdentifier.Currency.HISTORIC;
+            case CURRENT_MBI:
+                return PatientIdentifier.Currency.CURRENT;
+            default:
+                return PatientIdentifier.Currency.UNKNOWN;
         }
-        return system.equalsIgnoreCase(BENEFICIARY_ID);
     }
 
-    private static boolean isMatchingMbi(ICompositeType identifier, String historic) {
+    /**
+     * Get the currency information for CARIN standard MBI identifier
+     *
+     * @param identifier - the FHIR identifier
+     * @return the currency information
+     */
+    private static PatientIdentifier.Currency getCurrencyMbiStandard(ICompositeType identifier) {
```

Comment:
Honestly, this looks like a mess. Is there any way to get around so many nested if statements or refactor this?

Reply:
This is partially a product of still uncertain format and at times, duplication of this field in different places. I'm going to guess this will change.